### PR TITLE
@flexa/flexa one

### PR DESCRIPTION
**Pull request by ishitarastogi**
# Flexa

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] New Agent
- [ ] Fix an agent error
- [ ] Update an agent logic/test/documentation
- [ ] Code style update (formatting, renaming)
- [ ] Other (please describe): 

## New Agent PR readiness

- [x] `npm install` command works (i.e. package-lock is updated properly)
- [x] The `name` and `description` fields of `package.json` describe the agent properly
- [x] All the libraries listed in `package.json` are being used
- [x] All the development libraries are in `devDependencies` section of `package.json`
- [x] The agent README is updated
- [x] The agent README sufficiently describes what the agent does
- [x] All the alerts that the agent emits are listed in the README
- [ ] The txn/block examples listed in the README generate the expected findings
- [x] Code is easy to follow and/or properly comment
- [x] Agent contains tests
- [x] All tests pass when running `npm tests`
- [x] Tests are not making real network calls (i.e. network interaction is properly mocked)
- [x] Tests cover all the paths (i.e. negative & positive test cases)
- [x] When the tests complete, no warning/errors are shown
- [x] `npm start` works without errors when executed using a jsonRpcUrl for each of the networks specified in the README
- [x] The code do what the README describes
- [x] Network calls are properly being cached (if duplicated calls can occur)
- [x] A performance review has been done (e.g. `Promise.all` is used whenever possible)
- [x] The findings fields are filled out correctly
- [x] Findings `protocol` is specified (if the agent is for an specific one)
- [x] `metadata` of the findings contains all the information needed.
- [x] The SDK methods are being used appropriately (e.g. `filterLog`, `filterFunction`, `getEthersProvider`)
- [x] `ABI`s match the contracts used / `args` used from `LogDescriptions` exists in the `ABI`s 
- [x] the agent follows Forta recommended [best practices](https://docs.forta.network/en/latest/best-practices/)
- [x] Agent can recover from failed error call when they can fail (e.g. error are being catched on calls than can fail)

## Comments (optional)

If there is some useful information or important note for reviewing/understanding this agent please describe it here.

### CODE REVIEW

File: Flexa-Agents/Large-deposits/src/agent.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+          findings.push(newFinding);
+        }
+      }
+    }));
+
+    return findings;
+  };
+}
+
+
```
Comment (yasmine9) [inline Flexa-Agents/Large-deposits/src/agent.ts:100]: Please remove empty line here

File: Flexa-Agents/Large-deposits/src/agent.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      );
+
+      const isValidPartition = await flexaStakingContract.partitions(
+        decodedPartition,
+        { blockTag: txEvent.blockNumber },
+      );
+
+
+      if (isValidPartition) {
+        if (value.gte(amountThreshold)) {
```
Comment (yasmine9) [inline Flexa-Agents/Large-deposits/src/agent.ts:81]: can be merged into the same `if` statement

Comment (yasmine9) [inline Flexa-Agents/Large-deposits/src/agent.ts:81]: We are using 1m USD as a threshold, consider fetching the AMP token price in order to convert `value` to USD before comparing it to the threshold.

File: Flexa-Agents/Large-deposits/src/agent.spec.ts [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+  name: "Large Deposit",
+  description: "Large Deposit into staking pool",
+  alertId: "FLEXA-2",
+  severity: FindingSeverity.Info,
+  type: FindingType.Info,
+  metadata: {
+    value, fromPartition,
+    operator, from,
+    destinationPartition,
+    to, operatorData,
```
Comment (Kalzak) [inline Flexa-Agents/Large-deposits/src/agent.spec.ts:38]: For readability I'd recommend using the same formatting as you have in the `agent.ts` file where each part of the metadata has a key. See [here](https://github.com/NethermindEth/Forta-Agents/blob/548867fb5911110157e606f3c304d12bc99c9126/Flexa-Agents/Large-deposits/src/agent.ts#L31-L39).

File: Flexa-Agents/Large-deposits/README.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    * Severity is always set to "Info".
+    * Type is always set to "Info".
+    * Metadata contains: 
+        * `amountThreshold`: Threshold of 1M USD.
+        * `value`: The amount of tokens transferred.
+        * `fromPartition`: The partition the tokens were transfered from.
+        * `operator`: The address that initiated the transfer.
+        * `from`: The address the tokens were transferred from.
+        * `destinationPartition`: The partition the tokens were transferred to.
+        * `to`: The address the tokens were transferred to.
```
Comment (Kalzak) [inline Flexa-Agents/Large-deposits/README.md:24]: The metadata described here in the `README` is different to the metadata that is used in `agent.ts`.
See [here](https://github.com/NethermindEth/Forta-Agents/blob/548867fb5911110157e606f3c304d12bc99c9126/Flexa-Agents/Large-deposits/src/agent.ts#L31-L39).

File: Flexa-Agents/Large-deposits/src/agent.ts [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    severity: FindingSeverity.Info,
+    type: FindingType.Info,
+    metadata: {
+      value: amount.toString(),
+      fromPartition: partition.toLowerCase(),
+      operator: operator.toLowerCase(),
+      from: from.toLowerCase(),
+      destinationPartition: destinationPartition.toLowerCase(),
+      to: to.toLowerCase(),
+      operatorData: operatorData.toLowerCase(),
```
Comment (Kalzak) [inline Flexa-Agents/Large-deposits/src/agent.ts:38]: The metadata described here in `agent.ts` is different to the metadata described in `README.md`.
See [here](https://github.com/NethermindEth/Forta-Agents/blob/548867fb5911110157e606f3c304d12bc99c9126/Flexa-Agents/Large-deposits/README.md?plain=1#L18-L24).

**Review by yasmine9 — CHANGES_REQUESTED** [review]

File: Flexa-Agents/Large-deposits/src/agent.spec.ts [hunk 6]
```diff
@@ -0,0 +1,253 @@
+import { Finding, FindingSeverity, FindingType, HandleTransaction, TransactionEvent } from "forta-agent";
+import {
+  createAddress,
+  TestTransactionEvent,
+  encodeParameters,
+  MockEthersProvider,
+} from "forta-agent-tools";
```
Comment (Kalzak) [inline Flexa-Agents/Large-deposits/src/agent.spec.ts:7]: Formatting for large imports is inconsistent. I'd recommend a newline for each one when over three. This is just a recommendation for easy readability.

File: Flexa-Agents/Large-deposits/README.md [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+This agent detects large deposits to `FlexaCollateralManager` contract partitions.
+
+## Supported Chains
+
+- Ethereum
+
+## Alerts
+
+- FLEXA-2 
+    * Fired when it detects `TransferByPartition` events emitted from the `AMP` contract with a large amount transferred to a `FlexaCollaterManager` partition.
```
[the remark this answers is not in the document]
Comment (e1Ru1o) [reply Flexa-Agents/Large-deposits/README.md:14]: I was the one that changed that, I think your agent is already using flexa-1, so let's just skip this, and we can update all the alerts of the agents before delivering

File: Flexa-Agents/Large-deposits/src/agent.ts (hunk 2 above)
Comment (e1Ru1o) [reply Flexa-Agents/Large-deposits/src/agent.ts:81]: Good catch, I missed this one

**Review by Kalzak — CHANGES_REQUESTED** [review]

File: Flexa-Agents/Large-deposits/src/abi.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
+const AMP_TOKEN = [
+  `event TransferByPartition(
+    bytes32 indexed fromPartition,
+    address operator, 
+    address indexed from,
+    address indexed to,
+    uint256 value,
+    bytes data, 
+    bytes operatorData
+  )`,
```
[the remark this answers is not in the document]
Comment (e1Ru1o) [reply Flexa-Agents/Large-deposits/src/abi.ts:10]: No, that is used in this case to allow multi line strings

**Review by RCantu92 — APPROVED** [review]

File: Flexa-Agents/Large-deposits/src/abi.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (Kalzak) [reply Flexa-Agents/Large-deposits/src/abi.ts]: Oh I see, I might start using that myself actually!
My mistake, sorry.

**Pull request merged**